apps/prairielearn/assets/scripts/behaviors/react-fragments/index.tsx

```diff
@@ -0,0 +1,34 @@
+import { observe } from 'selector-observer';
+
+import { onDocumentReady } from '@prairielearn/browser-utils';
+import { type ComponentType, hydrate } from '@prairielearn/preact-cjs';
+
+import { ReactFragmentsRegistry } from './registry.js';
+
+const registry = new ReactFragmentsRegistry();
+
+export function registerReactFragment(id: string, component: ComponentType<any>) {
+  registry.setReactFragment(id, component);
+}
+
+onDocumentReady(() => {
+  observe('.js-react-fragment', {
+    async add(el) {
+      const id = el.id;
+      if (!id) throw new Error('js-react-fragment element must have an id');
+
+      // TODO: use a separate `data-component` attribute instead of the id?
```

Comment:
👀 let's consider addressing this. The idea was that the wrapper component would look something like this:

```html
<div class="js-react-fragment" data-component="InstructorInstanceAdminBillingForm">
```

It seems more explicit than using the ID for this.

Reply:
I addressed this with a randomId in [f603ed7](https://github.com/PrairieLearn/PrairieLearn/pull/12158/commits/f603ed76501071160ad1a78af2d0b4c1777560f7).